pennylane/estimator/estimate.py

```diff
@@ -405,6 +405,68 @@ def _resources_from_pl_ops(
     )
 
 
+def _get_symbolic_resource_decomposition(decomp_func, params, filtered_kwargs):
+    """Get resource decomposition for symbolic operators."""
+    base_cmpr_op = params.pop("base_cmpr_op")
+
+    target_params = base_cmpr_op.params.copy()
+    for k, v in filtered_kwargs.items():
+        if k in target_params and target_params[k] is None:
+            target_params[k] = v
+
+    params["target_resource_params"] = target_params
+
+    try:
+        return decomp_func(**params)
+    except TypeError as e:
+        msg = str(e)
+        if "unexpected keyword argument 'target_resource_params'" in msg:
+            raise TypeError(
+                f"The custom decomposition function '{decomp_func.__name__}' must accept "
+                "'target_resource_params' as a keyword argument."
+            ) from e
+
+        if (
+            "unexpected keyword argument 'num_ctrl_wires'" in msg
+            or "unexpected keyword argument 'num_zero_ctrl'" in msg
+        ):
+            raise TypeError(
+                f"The custom decomposition function '{decomp_func.__name__}' must accept "
+                "'num_ctrl_wires', 'num_zero_ctrl' and 'target_resource_params' as keyword arguments."
+            ) from e
+
+        if "unexpected keyword argument 'pow_z'" in msg:
+            raise TypeError(
+                f"The custom decomposition function '{decomp_func.__name__}' must accept "
+                "'pow_z' and 'target_resource_params' as keyword arguments."
+            ) from e
+        raise e
+
+
+def _get_resource_decomposition(comp_res_op: CompressedResourceOp, config: ResourceConfig):
+    """Get the resource decomposition for a compressed resource operator.
+
+    Args:
+        comp_res_op (:class:`~.pennylane.estimator.resource_operator.CompressedResourceOp`): operator in compressed representation to extract resources from
+        config (ResourceConfig | None):  A ``ResourceConfig`` object containing the additional
+        parameters required to estimate the resources for an operator. Defaults to
+        :class:`pennylane.estimator.resource_config.ResourceConfig`.
+
+    Returns:
+        list: The resource decomposition.
+    """
+    decomp_func, kwargs = _get_decomposition_function_and_kwargs(comp_res_op, config)
+
+    params = {key: value for key, value in comp_res_op.params.items() if value is not None}
+    filtered_kwargs = {key: value for key, value in kwargs.items() if key not in params}
+
+    op_type = comp_res_op.op_type
+    if op_type in (Adjoint, Controlled, Pow):
+        return _get_symbolic_resource_decomposition(decomp_func, params, filtered_kwargs)
```

Comment:
The logic is quite hard to follow here because you are handling two cases at once. This is a suggestion just to improve code clarity: 
```suggestion
    decomp_func, kwargs = _get_decomposition_function_and_kwargs(comp_res_op, config)

    op_type = comp_res_op.op_type
    if op_type in (Adjoint, Controlled, Pow):
        return _get_symbolic_resource_decomposition(decomp_func, comp_res_op.params, kwargs)

    params = {key: value for key, value in comp_res_op.params.items() if value is not None}
    filtered_kwargs = {key: value for key, value in kwargs.items() if key not in params}
```